contracts/oracle/AaveOracle.sol

```diff
@@ -0,0 +1,193 @@
+   *        aggregator is not consistent
+   * @param baseCurrency The base currency used for the price quotes. If USD is used, base currency is 0x0
+   * @param baseCurrencyUnit The unit of the base currency
+   */
+  constructor(
+    IPoolAddressesProvider provider,
+    address[] memory assets,
+    address[] memory sources,
+    address fallbackOracle,
+    address baseCurrency,
+    uint256 baseCurrencyUnit,
+    uint256 gracePeriod
+  ) {
+    ADDRESSES_PROVIDER = provider;
+    _setFallbackOracle(fallbackOracle);
+    _setAssetsSources(assets, sources);
+    BASE_CURRENCY = baseCurrency;
+    BASE_CURRENCY_UNIT = baseCurrencyUnit;
+    _gracePeriod = gracePeriod;
+    emit BaseCurrencySet(baseCurrency, baseCurrencyUnit);
+  }
+
+  /// @inheritdoc IAaveOracle
+  function setAssetSources(
+    address[] calldata assets,
+    address[] calldata sources
+  ) external override onlyAssetListingOrPoolAdmins {
+    _setAssetsSources(assets, sources);
+  }
+
+  /// @inheritdoc IAaveOracle
+  function setFallbackOracle(
+    address fallbackOracle
+  ) external override onlyAssetListingOrPoolAdmins {
+    _setFallbackOracle(fallbackOracle);
+  }
+
+  /**
+   * @notice Internal function to set the sources for each asset
+   * @param assets The addresses of the assets
+   * @param sources The address of the source of each asset
+   */
+  function _setAssetsSources(address[] memory assets, address[] memory sources) internal {
+    require(assets.length == sources.length, Errors.INCONSISTENT_PARAMS_LENGTH);
+    for (uint256 i = 0; i < assets.length; i++) {
+      assetsSources[assets[i]] = AggregatorInterface(sources[i]);
+      emit AssetSourceUpdated(assets[i], sources[i]);
+    }
+  }
+
+  /**
+   * @notice Internal function to set the fallback oracle
+   * @param fallbackOracle The address of the fallback oracle
+   */
+  function _setFallbackOracle(address fallbackOracle) internal {
+    _fallbackOracle = IPriceOracleGetter(fallbackOracle);
+    emit FallbackOracleUpdated(fallbackOracle);
+  }
+
+  /// @inheritdoc IPriceOracleGetter
+  function getAssetPrice(address asset) public view override returns (uint256) {
+    AggregatorInterface source = assetsSources[asset];
+
+    if (asset == BASE_CURRENCY) {
+      return BASE_CURRENCY_UNIT;
+    } else if (address(source) == address(0)) {
+      return _fallbackOracle.getAssetPrice(asset);
+    } else {
+      int256 price = source.latestAnswer();
+
+      bool stale = _isUpAndGracePeriodPassed(asset);
+      if (stale) {
+        revert(Errors.AGGREGATOR_IS_STALE);
+      }
+
+      if (price > 0) {
+        return uint256(price);
+      } else {
+        return _fallbackOracle.getAssetPrice(asset);
+      }
+    }
+  }
+
+  /// @inheritdoc IAaveOracle
+  function getAssetsPrices(
+    address[] calldata assets
+  ) external view override returns (uint256[] memory) {
+    uint256[] memory prices = new uint256[](assets.length);
```

Comment:
If there are concerns about gas costs I'd consider  limiting the number of assets that can be queried at once, also consider ways to reduce redundant calls if asset prices do not change often
maybe  =>    require(assets.length <= 100, "Limit of 100 assets exceeded");